plutus-contract/src/Plutus/Contract/Wallet.hs

```diff
@@ -165,18 +165,22 @@ balanceTx utxo pk UnbalancedTx{unBalancedTxTx} = do
         addCollateral utxo remainingFees tx''
 
 
--- | @addInputs mp pk vl tx@ selects transaction outputs worth at least
---   @vl@ from the UTXO map @mp@ and adds them as inputs to @tx@. A public
---   key output for @pk@ is added containing any leftover change.
+-- | @addInputs mp iv pk vl tx@ selects transaction outputs worth at least
+--   @vl@ from the UTXO map @mp@, with filtered out values from @iv@ that are already presented in unbalanced tx,
+--   and adds them as inputs to @tx@. A public key output for @pk@ is added containing any leftover change.
 addInputs
     :: Member (Error WalletAPIError) effs
     => UtxoMap
+    -> [Value]
```

Comment:
I think we should filter on `TxOutRef` (get the `TxOutRefs` from the inputs of the unbalanced transaction, and then use something like `Map.withoutKeys` on the utxo map). Using the `Value` of an output to identify the output seems quite brittle.

And then we might as well filter `mp` before calling `addInputs`

Reply:
Fixed 👌